text/0025-configuration-policy.md

```diff
@@ -0,0 +1,197 @@
+
+```ts
+// Base-line Stack to act as a subclass for all other Stacks.
+export class FooBarCompanyStack extends Stack {
+  public onConstructCreation(className: string, id: string, props: any): any {
+    if (className == 'aws-s3.Bucket') {
+      const bucketProps = <BucketProps>props;
+      if (bucketProps.encryption === undefined || bucketProps.encryption === BucketEncryption.UNENCRYPTED) {
+        return {
+          encryption: BucketEncryption.S3_MANAGED,
+          ...props,
+        };
+      }
+    }
+    return props;
+  }
+}
+// App-specific Stack which receives the defaults.
+export class MyAppStack extends FooBarCompanyStack {
+    constructor(scope: Construct, id: string, props?: StackProps) {
+        super(scope, id, props);
+        new Bucket(this, 'myBucket'); // Will automatically get S3-managed encryption.
+  }
+}
+```
+
+In addition to modifying the properties of the `Construct` being created, the
+`onConstructCreation` hook can be used to create entirely new resources. This
+example enforces a convention of each SQS Queue having its own dead-letter
+queue.
+
+```ts
+export class FooBarCompanyStack extends Stack {
+  public onConstructCreation(className: string, id: string, props: any): any {
+    if (className === 'aws-sqs.Queue') {
+      const queueProps = <QueueProps>props;
+      if (!id.includes("DLQ") && queueProps.deadLetterQueue === undefined) {
+        const dlq = new Queue(this, id + "DLQ");
+        return {
+          deadLetterQueue: { queue: dlq, maxReceiveCount: 3 },
+          ...props,
+        };
+      }
+    }
+    return props;
+  }
+}
+```
+
+# Motivation
+
+Teams, organizations, and companies may have need to enforce or apply default
+best practices to CDK-managed infrastructure. This approach will provide a
+consistent mechanism to apply these practices, while keeping the actual
+business-logic-containing Stacks lightweight.
+
+See the discussion in <https://github.com/aws/aws-cdk/issues/3235> for more
+context.
+
+# Basic Example
+
+_See README section._
+
+# Design Summary
+
+The approach adds a new hook (`onConstructCreation`) to the `Stack` class,
+which will be called by every L2 on construction. This hook receives the same
+inputs as the Construct's constructor, and enables users to adjust or override
+any of the input props, as well as throw errors or create ancillary resources.
+
+# Detailed Design
+
+A method is added to the `Stack` class with the following signature and default
+implementation:
+
+```ts
+public onConstructCreation(className: string, id: string, props: any): any {
+  return props;
+}
+```
+
+The `className` parameter will take the form of `module:class` (e.g.,
+`aws-s3.Bucket` or `aws-sqs.Queue`). The `id` and `props` parameters will be
+identical to the same parameters passed into the `Construct` on creation.
+
+All L2s will need to call this hook as part of their object construction. The
+call should be made immediately after the call to `super`, prior to property
```

Comment:
I think should this be called *before* the super call to allow maximum customizability. Otherwise, you won't be able to, e.g., customize the physical name of a L2.